rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1608 @@
+                // Remove it from our dispatcher
+                if self
+                    .device_update_notification_dispatcher
+                    .unregister_receiver(&receiver_id)
+                {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` removed from observed list"
+                    );
+                } else {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` not in observed list"
+                    );
+                }
+                Ok(())
+            }
+            adr_name_gen::NotificationPreferenceResponse::Failed => {
+                Err(Error(ErrorKind::ObservationError))
+            }
+        }
+    }
+
+    /// Hashes the device name and inbound endpoint name to create a single string.
+    fn hash_device_endpoint(device_name: &str, inbound_endpoint_name: &str) -> String {
+        // `~`` can't be in a topic token, so this will never collide with another device + inbound endpoint name combo
+        format!("{device_name}~{inbound_endpoint_name}")
+    }
+
+    /// Unhashes the device name and inbound endpoint name from a single string.
+    fn unhash_device_endpoint(hashed_device_endpoint: &str) -> Option<(String, String)> {
+        hashed_device_endpoint
+            .split_once('~')
+            .map(|(device_name, inbound_endpoint_name)| {
+                (device_name.to_string(), inbound_endpoint_name.to_string())
+            })
+    }
+
+    // ~~~~~~~~~~~~~~~~~ Asset APIs ~~~~~~~~~~~~~~~~~~~~~
+
+    /// Retrieves an [`Asset`] from the Azure Device Registry service.
+    ///
+    /// # Arguments
+    /// * `device_name` - The name of the device.
+    /// * `inbound_endpoint_name` - The name of the inbound endpoint.
+    /// * `asset_name` - The name of the asset.
+    /// * `timeout` - The duration until the client stops waiting for a response to the request, it is rounded up to the nearest second.
+    ///
+    /// Returns an [`Asset`] if the the asset was found.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidRequestArgument`](ErrorKind::InvalidRequestArgument)
+    /// if timeout is 0 or > `u32::max`.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if:
+    /// - device or inbound endpoint names are invalid.
+    /// - there are any underlying errors from the AIO RPC protocol.
+    ///
+    /// [`struct@Error`] of kind [`ValidationError`](ErrorKind::ValidationError)
+    /// if the asset name is empty.
+    pub async fn get_asset(
+        &self,
+        device_name: String,
+        inbound_endpoint_name: String,
+        asset_name: String,
+        timeout: Duration,
+    ) -> Result<Asset, Error> {
+        if asset_name.trim().is_empty() {
+            return Err(ErrorKind::ValidationError("asset_name".to_string()).into());
```

Comment:
```suggestion
            Err(Error(ErrorKind::ValidationError("asset_name".to_string()))
```
you don't need the return and ;, and then I think you should be able to do this to remove the into()